includes/cli.php

```diff
@@ -0,0 +1,408 @@
+ * https://github.com/Automattic/WPCOM-Legacy-Redirector
+ *
+ * For familiarity reasons, these commands work the same as those in that plugin.
+ * However our "redirect from" URLs need to be relative to the root of the site,
+ * and have a leading slash.
+ *
+ * @package hm-redirects
+ */
+
+namespace HM\Redirects\CLI;
+
+use WP_CLI, WP_CLI_Command;
+use HM\Redirects\{Utilities, Handle_Redirects};
+use const HM\Redirects\Post_Type\SLUG as REDIRECTS_POST_TYPE;
+
+/**
+ * Handles redirects in a scalable manner.
+ */
+class Commands extends WP_CLI_Command {
+
+	/**
+	 * Find domains redirected to.
+	 *
+	 * This is useful to populate WordPress' `allowed_redirect_hosts` filter.
+	 *
+	 * @subcommand find-domains
+	 *
+	 * @param string[] $args Not used.
+	 * @param string[] $assoc_args Not used.
+	 */
+	public function find_domains( array $args, array $assoc_args ) {
+		global $wpdb;
+
+		$domains        = array();
+		$paged          = 0;
+		$posts_per_page = 500;
+
+		$total_redirects = (int) $wpdb->get_var(
+			$wpdb->prepare(
+				"SELECT COUNT( ID ) FROM $wpdb->posts WHERE post_type = %s AND post_excerpt LIKE %s",
+				REDIRECTS_POST_TYPE,
+				'http%'
+			)
+		);
+
+		$progress_bar = WP_CLI\Utils\make_progress_bar( 'Finding domains', $total_redirects );
+
+		do {
+			$redirect_urls = $wpdb->get_col(
+				$wpdb->prepare(
+					"SELECT post_excerpt FROM $wpdb->posts WHERE post_type = %s AND post_excerpt LIKE %s ORDER BY ID ASC LIMIT %d, %d",
+					REDIRECTS_POST_TYPE,
+					'http%',
+					( $paged * $posts_per_page ),
+					$posts_per_page
+				)
+			);
+
+			foreach ( $redirect_urls as $redirect_url ) {
+				$progress_bar->tick();
+
+				if ( ! empty( $redirect_url ) ) {
+					$redirect_host = parse_url( $redirect_url, PHP_URL_HOST );
+					if ( $redirect_host ) {
+						$domains[] = $redirect_host;
+					}
+				}
+			}
+
+			// Throttle.
+			sleep( 1 );
+			$paged++;
+		} while ( count( $redirect_urls ) );
+
+		$progress_bar->finish();
+
+		$domains = array_unique( $domains );
+		WP_CLI::line( sprintf( 'Found %s unique outbound domains', number_format( count( $domains ) ) ) );
+
+		foreach ( $domains as $domain ) {
+			WP_CLI::line( $domain );
+		}
+	}
+
+	/**
+	 * Insert a single redirect
+	 *
+	 * from_url_relative must be relative to the root of the site.
+	 *
+	 * @subcommand insert-redirect
+	 * @synopsis <from_url_relative> <to_url_absolute>
+	 *
+	 * @param string[] $args Positional arguments.
+	 * @param string[] $assoc_args Not used.
+	 */
+	public function insert_redirect( $args, $assoc_args ) {
+		$from        = $args[0];
+		$status_code = 301;
+
+		if ( ctype_digit( $args[1] ) ) {
+			$to = get_permalink( (int) $args[1] );
+			if ( ! $to ) {
+				WP_CLI::error( sprintf( 'Destination post %s cannot be found', $args[1] ) );
+			}
+
+		} else {
+			$to = esc_url_raw( $args[1] );
+		}
+
+		$inserted = Utilities\insert_redirect( compact( 'from', 'to', 'status_code' ) );
```

Comment:
Not a fan of `compact()` at all, to me it's the same as `extract()`, only with slightly less voodoo.

Reply:
You can see what `compact()` does, and what it uses, just by looking at it. You called out `extract()` which is agreeably bad, but `compact()` does not invisibly import variables into scope like `extract()` does -- that's definitely voodoo.

It can't sensibly be used often, but here I think it has a succinctness in the way that a nice ternary operator statement sometimes has.